feat: add rocksdb write_options to configuration

tufitko opened the pull request:
#878
also, I've updated Dockerfile with Prerequisite command, because it didn't build
C++ isn't my common lang, so any comments are welcome

tufitko committed:
feat: added disableWAL option

tanruixiang commented:
May I ask why change the storage interface. I think it is necessary to reserve this `WriteOptions` parameter. Not all types share the same storage's writeoption(Even though all are now using the same defaults, I think it's still necessary to keep).

tufitko commented:
@tanruixiang I can keep old Write interface, I've removed it because I didn't see non-default write options. Can you give an example of when this might be needed?

tanruixiang commented:
> @tanruixiang I can keep old Write interface, I've removed it because I didn't see non-default write options. Can you give an example of when this might be needed?

There are many options in `WriteOptions`, not just `disableWAL`. If someone needs to change other options like you, and use different options in different places, they will need this `WriteOptions` parameter. In particular, the `rocksdb::WriteOptions write_opts_ ` you add are private, which means we can't make changes in places where options might need to be changed.

tufitko commented:
🤔 @tanruixiang  I think I need to return the old Write interface and add a method to return default write options from storage. Is it ok?

mapleFU commented:
Hi, I think we can provide a `defaultRocksDBOption` or `writeWithDefault` in `storage`, I think just create default `WriteOptions` everywhere is not a good idea, and only those who know what rocksdb's options means can define them themselves

tanruixiang commented:
> 🤔 @tanruixiang I think I need to return the old Write interface and add a method to return default write options from storage. Is it ok?

We can discuss it. @git-hulk @PragmaTwice What form do you think is the best?

PragmaTwice commented on Dockerfile:
```diff
@@ -15,20 +15,20 @@
 
 RUN apt update
-RUN apt install -y cmake make git autoconf libtool g++ python3
+RUN apt install -y git gcc g++ make cmake autoconf automake libtool python3 libssl-dev
```
I pick this line to #887, since this PR may not be merged quickly. Thanks for your contribution!

tufitko replied:
ok! Then I remove it from my PR

git-hulk commented:
I also prefer to add the `WriteWithDefault` to avoid creating the option everywhere, then we can keep the `Write` method to be consistent with RocksDB.

tufitko commented:
So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.

PragmaTwice commented:
> So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.

I think you can recover the WriteOptions parameter, and just expose a method `WriteOptions Storage::DefaultWriteOptions()`.

And you can also add something like `DeleteWithDefault(...) { Delete(DefaultWriteOptions(), ...) }` if you like.

tufitko commented:
> > So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.
> 
> I think you can recover the WriteOptions parameter, and just expose a method `WriteOptions Storage::DefaultWriteOptions()`.
> 
> And you can also add something like `DeleteWithDefault(...) { Delete(DefaultWriteOptions(), ...) }` if you like.

reworked again 🙈

PragmaTwice commented on src/storage.cc:
```diff
@@ -511,7 +512,7 @@ rocksdb::Status Storage::Write(const rocksdb::WriteOptions &options, rocksdb::Wr
   }
 
-  return db_->Write(options, updates);
+  return db_->Write(write_opts_, updates);
```
```suggestion
  return db_->Write(options, updates);
```

tufitko replied:
ooops

tufitko replied:
fixed

PragmaTwice commented on src/config.cc:
```diff
@@ -174,6 +174,7 @@ Config::Config() {
       {"rocksdb.delayed_write_rate", false, new Int64Field(&RocksDB.delayed_write_rate, 0, 0, INT64_MAX)},
       {"rocksdb.wal_ttl_seconds", true, new IntField(&RocksDB.WAL_ttl_seconds, 3*3600, 0, INT_MAX)},
       {"rocksdb.wal_size_limit_mb", true, new IntField(&RocksDB.WAL_size_limit_MB, 16384, 0, INT_MAX)},
+      {"rocksdb.disableWAL", true, new YesNoField(&RocksDB.disableWAL, false)},
```
I think `disable_wal` is more suitable in configuration file? since other options are all underscore-case and case-insensitive.

tufitko replied:
`disable_wal` looks better, but in RocksDb it calls `disableWAL`. shouldn't we save the original name?

git-hulk replied:
I also found this point, I think it should be inconsistent in RocksDB NOT the intent. `disable_wal` even `disable_WAL` is better than `disableWAL`, what do you think?

tanruixiang commented:
As @PragmaTwice  said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.

tufitko commented:
> As @PragmaTwice said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.

hmmm, I can add `sync`, `no_slowdown`, `low_pri`, `memtable_insert_hint_per_batch`. But `ignore_missing_column_families` and `timestamp` looks useless to setting through the config. do you agree with this?

tanruixiang commented:
> > As @PragmaTwice said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.
> 
> hmmm, I can add `sync`, `no_slowdown`, `low_pri`, `memtable_insert_hint_per_batch`. But `ignore_missing_column_families` and `timestamp` looks useless to setting through the config. do you agree with this?

I think these parameters can be reserved, and synchronization in the config file(For example, adding an interface to change the `default writeoption`. If the configuration file is configured with `default writeoption`, then make corresponding modifications.). Of course, if it is not in the configuration file, the default value is used.

tanruixiang commented:
> I think these parameters can be reserved, and synchronization in the config file(For example, adding an interface to change the `default writeoption`. If the configuration file is configured with `default writeoption`, then make corresponding modifications.). Of course, if it is not in the configuration file, the default value is used.

After this modification, if you want to `disableWAL`, you only need to modify the configuration file.

tufitko commented:
@tanruixiang updated, I'm not sure I understood correctly. 
I didn't add `timestamp` because I still think it's useless. also, there isn't it in newer versions of rocksdb.

PragmaTwice commented:
I prefer to expose only the options that may be used in practice. If we increase the number of options exposed, then we need to consider how kvrocks will behave if these options are used, and whether all aspects of the functionality in kvrocks will still work. 

If we can't be sure of that, then just exposing `disbale_wal` is fine. If you expose other options, you may need to do some additional explaining: such as whether you tested how kvrocks works when those options are used, and under what circumstances they will be used.

BTW, do not forget to update `kvrocks.conf` : )

tufitko commented:
I need a decision about exposing write options. 
I think `ignore_missing_column_families` and `timestamp` are unneeded for configuring. Others can be useful

tanruixiang commented:
There are some parameter settings in writeoption that should be safe, such as `sync`. I think it's okay to keep these parameters, and if the user uses a parameter in config file that we're not sure about, then we give a warning (I think for users unfamiliar with writeoption, they'll just use the default values, and users who modify those values are sure Is to have a certain understanding of rocksdb).

PragmaTwice reviewed (dismissed):
LGTM

git-hulk commented on src/config.cc:
```diff
@@ -603,6 +611,9 @@ Status Config::finish() {
   }
+  if (!slaveof_.empty() && RocksDB.write_options.disable_WAL) {
+    return Status(Status::NotOK, "cannot disable the WAL while using Master-Slave replication");
+  }
```
I think we should check this in the `slaveof` command.

tufitko replied:
ok, moved to `Execution` of slaveof

git-hulk replied:
Thank you!

git-hulk reviewed (commented):
Others are good to me except one comment.
